tests/test_main.py

```diff
@@ -322,3 +322,34 @@ def test_packages_distributions_neither_toplevel_nor_files(self):
             prefix=self.site_dir,
         )
         packages_distributions()
+
+    def test_packages_distributions_all_module_types(self):
+        """
+        Test top-level modules detected on a package without 'top-level.txt'.
+        """
+        suffixes = importlib.machinery.all_suffixes()
+        fixtures.build_files(
+            {
+                'all_distributions-1.0.0.dist-info': {
+                    'METADATA': """
+                        Name: all_distributions
+                        Version: 1.0.0
+                    """,
+                    'RECORD': ''.join(
+                        f'{i}-top-level{suffix},,\n'
+                        f'{i}-in-namespace/mod{suffix},,\n'
+                        f'{i}-in-package/__init__.py,,\n'
+                        f'{i}-in-package/mod{suffix},,\n'
+                        for i, suffix in enumerate(suffixes)
```

Comment:
I'd prefer if this test were parametrized by `suffix` such that `{i}` would be unnecessary. Since we're reliant on unittest and don't have good parameterization support, let's go with this for now.